### PR TITLE
feat(EMS-4025-4045): declarations - modern slavery - data saving

### PR DESCRIPTION
**Pull request by ttbarnes**
## Introduction :pencil2:
This PR adds data saving to the new "Declarations - Modern slavery" form.

## Resolution :heavy_check_mark:
- Update SQL dump.
- Add E2E test.
- Update `ALLOWED_GRAPHQL_RESOLVERS`.
- Create new declaration / populated declaration helper functions.
- Create new UI GQL mutation for `updateDeclarationModernSlavery`.
- Update the Modern slavery POST controller to call a new `save.declarationModernSlavery` function.
- Update `flattenApplicationData` to return modern slavery fields.
- Update UI test mocks and types.
- Update nunjucks components.

## Miscellaneous :heavy_plus_sign:
- Remove the need to declare `declarations: declarations.update` in UI API/GQL queries structure.
- Fix `declarations` vs `declaration` typo.
- Add a missing type to the UI's `mockApplication` definition.
- Fix some typos in "company details" controller and simplify the validation directory structure.

### CODE REVIEW

File: database/exip.sql [hunk 1]
```diff
@@ -593,12 +593,13 @@ CREATE TABLE `Declaration` (
   `agreeToAntiBribery` tinyint(1) DEFAULT NULL,
   `agreeToConfidentiality` tinyint(1) DEFAULT NULL,
   `agreeToConfirmationAndAcknowledgements` tinyint(1) DEFAULT NULL,
-  `application` varchar(191) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
+  `application` varchar(191) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
```
Comment (Zainzzkk) [inline database/exip.sql:596]: should this have changed?

File: src/api/helpers/get-declaration-by-id/index.ts [hunk 2]
```diff
@@ -4,22 +4,31 @@ import { Context } from '.keystone/types'; // eslint-disable-line
...
  * @returns {Promise<ApplicationDeclaration>}
  */
-const getDeclarationById = async (context: Context, id: string) => {
+const getDeclarationById = async (context: Context, declarationId: string) => {
   try {
-    console.info('Getting declaration by ID %s', id);
+    console.info('Getting declaration by ID %s', declarationId);
 
+    // TODO: unit test.
+    // TODO: new helper.
```
Comment (Zainzzkk) [inline src/api/helpers/get-declaration-by-id/index.ts:15]: ticket numbers please

File: src/api/helpers/get-declaration-by-id/index.ts [hunk 3]
```diff
@@ -4,22 +4,31 @@ import { Context } from '.keystone/types'; // eslint-disable-line
...
     const declaration = await context.db.Declaration.findOne({
-      where: { id },
+      where: { id: declarationId },
     });
 
-    return declaration;
+    const declarationModernSlavery = await context.db.DeclarationModernSlavery.findOne({
+      where: { id: declaration?.modernSlaveryId },
```
Comment (Zainzzkk) [inline src/api/helpers/get-declaration-by-id/index.ts:21]: does this need to be optional chained?

File: src/ui/server/controllers/insurance/declarations/modern-slavery/index.test.ts [hunk 4]
```diff
@@ -179,11 +197,45 @@ describe('controllers/insurance/declarations/modern-slavery', () => {
...
+      it(`should redirect to ${PROBLEM_WITH_SERVICE}`, async () => {
+        await post(req, res);
 
         expect(res.redirect).toHaveBeenCalledWith(PROBLEM_WITH_SERVICE);
       });
     });
+
+    describe('api error handling', () => {
+      describe('save data call', () => {
+        describe('when the save data API call does not return anything', () => {
```
Comment (Zainzzkk) [inline src/ui/server/controllers/insurance/declarations/modern-slavery/index.test.ts:209]: should this say false or not return anything?

File: src/ui/server/controllers/insurance/declarations/save-data/modern-slavery/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  });
+
+  describe('when there is an error calling the API', () => {
+    beforeEach(() => {
+      updateApplicationSpy = mockSpyPromiseRejection;
+      api.keystone.application.update.declarationModernSlavery = updateApplicationSpy;
+    });
+
+    it('should throw an error', async () => {
+      try {
```
Comment (Zainzzkk) [inline src/ui/server/controllers/insurance/declarations/save-data/modern-slavery/index.test.ts:57]: .toThrows?

File: src/api/helpers/get-declaration-by-id/index.ts (hunk 2 above)
Comment (ttbarnes) [reply src/api/helpers/get-declaration-by-id/index.ts:15]: WIll be adressing in this PR.

File: src/ui/server/controllers/insurance/declarations/modern-slavery/index.test.ts (hunk 4 above)
Comment (ttbarnes) [reply src/ui/server/controllers/insurance/declarations/modern-slavery/index.test.ts:209]: Updated all instances of this.

File: src/api/helpers/get-declaration-by-id/index.ts (hunk 3 above)
Comment (ttbarnes) [reply src/api/helpers/get-declaration-by-id/index.ts:21]: Yes, `declaration` might not exist.

File: database/exip.sql (hunk 1 above)
Comment (ttbarnes) [reply database/exip.sql:596]: Yes, it now matches what a SQL dump export has. Seems to have been missed.

File: src/ui/server/controllers/insurance/declarations/save-data/modern-slavery/index.test.ts (hunk 5 above)
Comment (ttbarnes) [reply src/ui/server/controllers/insurance/declarations/save-data/modern-slavery/index.test.ts:57]: We have this throughout the codebase, would like to be consistent.

**Review by Zainzzkk — APPROVED** [review]

**Review by Zainzzkk — APPROVED** [review]
can you add something in the changes about moving the validation files? in company-details

File: src/ui/templates/partials/insurance/declarations/modern-slavery/has-no-offenses-or-investigations.njk [hunk 6]
```diff
@@ -6,6 +6,20 @@
 
 {% set fieldId = FIELDS.HAS_NO_OFFENSES_OR_INVESTIGATIONS.ID %}
 
+{#
```
[the remark this answers is not in the document]
Comment (ttbarnes) [reply src/ui/templates/partials/insurance/declarations/modern-slavery/has-no-offenses-or-investigations.njk:9]: This is in a partial?

If you mean for setting the variable, I don't think it's possible to declare a nunjucks variable in a partial and return it

**Comment by ttbarnes** [comment]
> can you add something in the changes about moving the validation files? in company-details

Updated

**Pull request merged**